### PR TITLE
XWIKI-22547: Make websocket work with root servlet context

### PR DESCRIPTION
**Pull request by promasu**
# Jira URL

https://jira.xwiki.org/projects/XWIKI/issues/XWIKI-22547

# Changes

## Description

* Add a check if the webapppath
is empty, as in set to `/` and exclude it from the URL building.

## Clarifications

* Exclude the edge-case
and keep the generic code.

# Screenshots & Video

# Executed Tests

New test case was added.
✅ Checked with upstream code without fix to detect this issue.
✅ Run test locally
✅ Deploy fixed module to formerly faulty XWiki instance and observe the issue being fixed.

# Expected merging strategy

* Prefers squash: Yes 
* Backport on branches:
  *

### CODE REVIEW

File: xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java [hunk 1]
```diff
@@ -84,9 +84,10 @@ public String url(String pathOrRoleHint)
             XWikiContext xcontext = this.xcontextProvider.get();
 
             StringBuilder path = new StringBuilder("/");
-            path.append(xcontext.getWiki().getWebAppPath(xcontext));
+            if (!xcontext.getWiki().getWebAppPath(xcontext).equals('/')) {
```
Comment (surli) [inline xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java:87]: At first look `getWebAppPath` cannot be null, but to be safer I'd use `!"/".equals(...)`.

Comment (promasu) [reply xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java:87]: Sure. I'll change that.
The current code breaks if the value of `getWebAppPath` would ever become null, doesn't it?

Comment (promasu) [reply xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java:87]: Sure. I'll change that.
The current code breaks if the value of `getWebAppPath` would ever become null, doesn't it?

Comment (surli) [reply xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java:87]: Yes right now if `getWebAppPath` returns null you'd always get a NullPointerException.

**Review by surli — APPROVED** [review]

File: xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java [hunk 2]
```diff
@@ -84,9 +84,13 @@ public String url(String pathOrRoleHint)
             XWikiContext xcontext = this.xcontextProvider.get();
 
             StringBuilder path = new StringBuilder("/");
-            path.append(xcontext.getWiki().getWebAppPath(xcontext));
-            path.append("websocket").append(path(pathOrRoleHint));
 
+            // Test if installed with short URL and avoid double slash
```
Comment (vmassol) [inline xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java:88]: This comment could be improve by mentioning XWiki installed under the ROOT servlet context (since Short URL is something different)

File: xwiki-platform-core/xwiki-platform-websocket/src/test/java/org/xwiki/websocket/script/WebSocketScriptServiceTest.java [hunk 3]
```diff
@@ -109,4 +109,11 @@ void urlWithException() throws Exception
             "Failed to create WebSocket URL for [test]. Root cause is [MalformedURLException: Invalid server URL!].",
             this.logCapture.getMessage(0));
     }
+
+    @Test
+    void urlWithEmptyWebAppPath() {
```
Comment (vmassol) [inline xwiki-platform-core/xwiki-platform-websocket/src/test/java/org/xwiki/websocket/script/WebSocketScriptServiceTest.java:114]: Rather than "empty", I think it's more correct to mention  "Root" servlet context.

File: xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java [hunk 4]
```diff
@@ -84,9 +84,13 @@ public String url(String pathOrRoleHint)
...
+            if (!"/".equals(xcontext.getWiki().getWebAppPath(xcontext))) {
+                path.append(xcontext.getWiki().getWebAppPath(xcontext));
+            }
```
Comment (vmassol) [inline xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java:91]: Instead of this, why not simply change the line above and remove this IF:

From:

```
StringBuilder path = new StringBuilder("/");
```

To:

```
StringBuilder path = new StringBuilder();
```

?

**Review by vmassol — CHANGES_REQUESTED** [review]

File: xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java (hunk 4 above)
Comment (vmassol) [reply xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java:91]: if `getWebAppPath()`  doesn't always return a path prefixed with `/` then just check for that if it doesn't start with `/` then add it.

Comment (vmassol) [reply xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java:91]: Actually just checked and indeed, `getWebAppPath()` never returns a leading "/", see https://github.com/xwiki/xwiki-platform/blob/67c0e919f2c3d0e4385b2133e7cc0318da598834/xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/XWiki.java#L5456

Comment (promasu) [reply xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java:91]: It does not return a leading "/" but always a trailing one, right?

My idea was to exclude this single edge case where the leading "/" is equal to the trailing one which leads to this code returning "//". Or do I get you wrong?

Comment (vmassol) [reply xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java:91]: @promasu thanks, what you did is good.

**Review by vmassol — APPROVED** [review]

File: xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java (hunk 4 above)
Comment (promasu) [reply xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java:91]: Ok perfect. Thanks!

File: xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java [hunk 5]
```diff
@@ -84,9 +84,13 @@ public String url(String pathOrRoleHint)
             XWikiContext xcontext = this.xcontextProvider.get();
 
             StringBuilder path = new StringBuilder("/");
-            path.append(xcontext.getWiki().getWebAppPath(xcontext));
-            path.append("websocket").append(path(pathOrRoleHint));
 
+            // Test if installed with ROOT servlet context and avoid double slash
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java:88]:
```suggestion
            // Test if installed with ROOT servlet context and avoid double slash.
```

File: xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java [hunk 6]
```diff
@@ -84,9 +84,13 @@ public String url(String pathOrRoleHint)
...
+            if (!"/".equals(xcontext.getWiki().getWebAppPath(xcontext))) {
+                path.append(xcontext.getWiki().getWebAppPath(xcontext));
+            }
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-websocket/src/main/java/org/xwiki/websocket/script/WebSocketScriptService.java:91]:
```suggestion
            String webAppPath = xcontext.getWiki().getWebAppPath(xcontext);
            if (!"/".equals(webAppPath)) {
                path.append(webAppPath);
            }
```

**Review by mflorea — COMMENTED** [review]
Looks good. I left two suggestions.

**Review by mflorea — APPROVED** [review]

**Comment by mflorea** [comment]
Thank you, @promasu, for taking the time to investigate and fix this issue. This is highly appreciated.
